src/Microsoft.Graph.Auth/ConfidentialClient/AuthorizationCodeProvider.cs

```diff
@@ -0,0 +1,94 @@
+        /// Creates a new <see cref="IConfidentialClientApplication"/>.
+        /// </summary>
+        /// <param name="clientId">Client ID (also known as <i>Application ID</i>) of the application as registered in the application registration portal (https://aka.ms/msal-net-register-app).</param>
+        /// <param name="redirectUri">Also named <i>Reply URI</i>, the redirect URI is the URI where the STS (Security Token Service) will call back the application
+        ///  with the security token. For details see https://aka.ms/msal-net-client-applications.</param>
+        /// <param name="clientCredential">A <see cref="Microsoft.Identity.Client.ClientCredential"/> created either from an application secret or a certificate.</param>
+        /// <param name="tokenStorageProvider">An <see cref="ITokenStorageProvider"/> for storing and retrieving access token. </param>
+        /// <param name="tenant">Tenant to sign-in users. This defaults to <c>common</c> if non is specified.</param>
+        /// <param name="nationalCloud">A <see cref="NationalCloud"/> which identifies the national cloud endpoint to use as the authority. This defaults to the global cloud <see cref="NationalCloud.Global"/> (https://login.microsoftonline.com).</param>
+        /// <returns>A <see cref="IConfidentialClientApplication"/></returns>
+        public static IConfidentialClientApplication CreateClientApplication(string clientId,
+            string redirectUri,
+            ClientCredential clientCredential,
+            ITokenStorageProvider tokenStorageProvider = null,
+            string tenant = null,
+            NationalCloud nationalCloud = NationalCloud.Global)
+        {
+            TokenCacheProvider tokenCacheProvider = new TokenCacheProvider(tokenStorageProvider);
+            string authority = NationalCloudHelpers.GetAuthority(nationalCloud, tenant ?? AuthConstants.Tenants.Common);
+            return new ConfidentialClientApplication(clientId, authority, redirectUri, clientCredential, tokenCacheProvider.GetTokenCacheInstnce(), null);
+        }
+
+        /// <summary>
+        /// Adds an authentication header to the incoming request by checking the application's <see cref="TokenCache"/>
+        /// for an unexpired access token.
+        /// If an access token doesn't exist, it will throw a <see cref="AuthenticationException"/>
+        /// and the web app must handle this and perform a challange.
+        /// </summary>
+        /// <param name="httpRequestMessage">A <see cref="HttpRequestMessage"/> to authenticate.</param>
+        public async Task AuthenticateRequestAsync(HttpRequestMessage httpRequestMessage)
+        {
+            MsalAuthenticationProviderOption msalAuthProviderOption = httpRequestMessage.GetMsalAuthProviderOption();
+            AuthenticationResult authenticationResult = await GetAccessTokenSilentAsync(msalAuthProviderOption);
+
+            if (string.IsNullOrEmpty(authenticationResult?.AccessToken))
```

Comment:
Just so I'm clear, this is the only provider that throws an AuthenticationException instead of requesting an access token because ASP.NET (and Core) need to handle this at the controller so that the user is directed to authenticate, correct?

Reply:
Correct.